### PR TITLE
experimental/swa_attention_cutile

### PR DESCRIPTION
**Pull request by DevTechJr**
## Description

Adds an experimental sliding window attention (SWA) prefill kernel. This operator is not currently in TileGym and is needed for native SWA support in Mistral, Mixtral, and Gemma 3.

### Kernel

Single fused cuTile kernel with online softmax. Each query attends to at most W preceding keys. The kernel skips KV blocks outside the window entirely, giving O(S*W) instead of O(S^2).

- 2D grid: `bid(0)` = Q tile block, `bid(1)` = batch * head
- exp2 + flush-to-zero softmax (maps to SFU hardware)
- Combined trailing window + causal + seq bounds mask in one pass
- GQA support via KV head expansion
- Decode falls back to PyTorch SDPA (this is a prefill kernel)

### Parameters

Autotuned on B300 SXM6 AC (sm_103) with two rounds:

1. In-house grid search over 54 tile/occupancy/precision configs, hardware-pruned to 42, quality-gated at cosine sim >= 0.99 vs fp32 reference, median of 200 CUDA event trials with top-3 re-verification at 500 trials
2. NVBench (cuda-bench) cold validation with L2 cache flushing, automatic noise detection, and GPU clock locking

Winner: `TILE_M=64 TILE_N=128 occupancy=2 precision=fast`

### Performance

SWA vs Flash Attention (PyTorch SDPA), B=1 H=32 D=128 W=4096, NVBench cold measurements:

| Seq | SWA (us) | SDPA (us) | Speedup |
|-----|----------|-----------|---------|
| 4096 | 258 | 461 | 1.78x |
| 8192 | 678 | 1571 | 2.32x |
| 16384 | 1542 | 5781 | 3.75x |
| 32768 | 3295 | 22100 | 6.71x |

Cosine similarity 1.00000 vs fp32 reference at all tested lengths. Noise < 1% on all NVBench runs.

### Files

New:
- `src/tilegym/ops/cutile/experimental/swa_attention.py` - kernel, host launcher, HF model integration
- `tests/ops/experimental/test_swa_attention.py` - 14 correctness tests (basic, edge cases, various shapes, 8K Mistral context)
- `tests/benchmark/experimental/bench_swa_attention.py` - Triton perf_report benchmark

Modified:
- `src/tilegym/ops/ops.py` - added `@dispatch("swa_attention")`
- `src/tilegym/ops/cutile/__init__.py` - added experimental imports and `__all__` entries

### HF model integration

Includes `apply_tilegym_swa_to_mistral()` which follows the same monkey-patch pattern as the existing `apply_tilegym_kernel_to_mistral` in `monkey_patch.py`. Replaces `ALL_ATTENTION_FUNCTIONS["sdpa"]` with an SWA-aware wrapper. Tested with Mistral-7B-Instruct-v0.3 end-to-end generation.

### What this unblocks

From the roadmap:
- Mixtral-8x7B (W=4096) - listed as "Help Wanted"
- Gemma 3 (variable W per layer) - WIP, needs SWA
- Mistral-7B - has E2E support but currently uses full causal instead of native SWA

## CI Configuration
```yaml
config:
  build: true
  # valid options are "ops", "benchmark", and "sanity"
  test: ["ops", "benchmark"]
```

## Checklist
- [x] Code formatted and imports sorted via repo specifications (`./format.sh`)
- [x] Documentation updated (if needed)
- [x] CI configuration reviewed

### CODE REVIEW

**Comment by hannahli-nv** [comment]
Hi @DevTechJr, thanks for the contribution! We have received your CLA file and will review your PR later.

**Comment by hannahli-nv** [comment]
/ok to test https://github.com/NVIDIA/TileGym/commit/3fe1482ba6b9cca9e504058f8e3eca852b14cae4

File: src/tilegym/ops/cutile/experimental/swa_attention.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        p = ct.exp2(qk * scale_log2 - ct.expand_dims(m_new, axis=1), flush_to_zero=True)
+
+        # update running sum and weighted accumulator
+        l_i = alpha * l_i + ct.sum(p, axis=1)
+        p_fp16 = ct.astype(p, ct.float16)  # downcast for the PV matmul
+        acc = ct.expand_dims(alpha, axis=1) * acc + ct.mma(p_fp16, v_tile, ct.zeros((TILE_M, TILE_D), dtype=ct.float32))
+        m_i = m_new
+
+    # final normalization: divide accumulated values by softmax denominator
+    out = acc / ct.expand_dims(l_i, axis=1)
```
Comment (azazhu) [inline src/tilegym/ops/cutile/experimental/swa_attention.py:114]: Is it possible that all values ​​of l_i are 0? If so, is it necessary to use ct.maximum to find a minimum number?

File: src/tilegym/ops/cutile/experimental/swa_attention.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        offs_n = ct.arange(TILE_N, dtype=ct.int32) + kv_start
+        offs_n_2d = ct.expand_dims(offs_n, axis=0)
+        offs_m_2d = ct.expand_dims(offs_m, axis=1)
+
+        mask = offs_n_2d > (offs_m_2d - window_size)  # trailing window
+        if CAUSAL:
+            mask = mask & (offs_n_2d <= offs_m_2d)  # causal (no future keys)
+        mask = mask & (offs_n_2d < seq_k)  # don't read past actual seq len
+
+        qk = ct.where(mask, qk, ct.full((TILE_M, TILE_N), -1e30, dtype=ct.float32))
```
Comment (azazhu) [inline src/tilegym/ops/cutile/experimental/swa_attention.py:99]: nit: define _NEG_INF = -1e30

File: tests/ops/experimental/test_swa_attention.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            swa_reference,
+            kwargs={"q": q, "k": k, "v": v, "window_size": W, "is_causal": True},
+            rtol=1e-2,
+            atol=5e-2,
+        )
+
+    # -- basic correctness --
+
+    @pytest.mark.parametrize("backend", _backends)
+    def test_window_equals_seq(self, backend):
```
Comment (azazhu) [inline tests/ops/experimental/test_swa_attention.py:63]: Change the test function to start with test_op_xxx

File: src/tilegym/ops/cutile/experimental/swa_attention.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+
+# -- host launcher --
+
+_DEFAULT_TILE_M = 64
+_DEFAULT_TILE_N = 128  # autotuned: 1.9x faster than N=64 on B300
+
+
+def tile_swa_attention(q, k, v, window_size, scaling=None, is_causal=True, **kwargs):
+    # q: (B, H, S_Q, D), k/v: (B, H_K, S_K, D) -- fp16
```
Comment (azazhu) [inline src/tilegym/ops/cutile/experimental/swa_attention.py:125]: nit: looks this func only support fp16. suggest check dtype at beginning:
```
if q.dtype not in (torch.float16,):
    raise ValueError(f"SWA kernel requires fp16 input, got {q.dtype}")
```

File: src/tilegym/ops/cutile/experimental/swa_attention.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    # these are optional -- attention is the main target.
+    try:
+        from tilegym.ops import get_apply_rope_func
+        from tilegym.ops import get_rms_norm_module
+        from tilegym.ops import get_swiglu_module
+
+        modeling_mistral.apply_rotary_pos_emb = get_apply_rope_func(model="llama")
+        modeling_mistral.MistralRMSNorm = get_rms_norm_module()
+        modeling_mistral.MistralMLP = get_swiglu_module()
+    except Exception:
```
Comment (azazhu) [inline src/tilegym/ops/cutile/experimental/swa_attention.py:273]: nit: 
except ImportError: pass
except Exception: raise

File: tests/ops/experimental/test_swa_attention.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    i = torch.arange(S_Q, device=q.device).unsqueeze(1)
+    j = torch.arange(S_K, device=q.device).unsqueeze(0)
+    mask = j > (i - window_size)
+    if is_causal:
+        mask = mask & (j <= i)
+    scores = scores.masked_fill(~mask.unsqueeze(0).unsqueeze(0), float("-inf"))
+    return torch.matmul(torch.softmax(scores, dim=-1), v.float()).to(q.dtype)
+
+
+class TestSWAAttention(common.PyTestCase):
```
Comment (azazhu) [inline tests/ops/experimental/test_swa_attention.py:34]: if this func supports GQA, pls add a test_op_gqa.

File: src/tilegym/ops/cutile/experimental/swa_attention.py (hunk 1 above)
Comment (DevTechJr) [reply src/tilegym/ops/cutile/experimental/swa_attention.py:118]: Thank you for pointing out the edge case. I added a check to handle this edge case gracefully.

File: src/tilegym/ops/cutile/experimental/swa_attention.py (hunk 2 above)
Comment (DevTechJr) [reply src/tilegym/ops/cutile/experimental/swa_attention.py]: Added this as a constant instead of hardcoding it.

File: tests/ops/experimental/test_swa_attention.py (hunk 3 above)
Comment (DevTechJr) [reply tests/ops/experimental/test_swa_attention.py]: Renamed function accordingly.

File: src/tilegym/ops/cutile/experimental/swa_attention.py (hunk 4 above)
Comment (DevTechJr) [reply src/tilegym/ops/cutile/experimental/swa_attention.py:129]: Thank you for the suggestion. Added the dtype check with ValueError for non-fp16 inputs.

**Comment by DevTechJr** [comment]
Hi @azazhu , thank you for the comments. I have resolved feedback from both you and Copilot in my new push. Please test and let me know next steps accordingly.

File: tests/benchmark/experimental/bench_swa_attention.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    scores = scores.masked_fill(~mask.unsqueeze(0).unsqueeze(0), float("-inf"))
+    return torch.matmul(torch.softmax(scores, dim=-1), v.float()).to(q.dtype)
+
+
+register_impl("swa_attention", "torch")(reference_swa)
+
+
+ALL_BACKENDS = [
+    ("cutile", "CuTile SWA", ("blue", "-")) if is_backend_available("cutile") else None,
+    ("torch", "PyTorch ref", ("green", "-")),
```
Comment (hannahli-nv) [inline tests/benchmark/experimental/bench_swa_attention.py:38]: Hi @DevTechJr , I notice that the display names here contain spaces ("CuTile SWA", "PyTorch ref"), which breaks the CI summary page. Please match the convention rename to single-token names:
```suggestion
    ("cutile", "CuTile", ("blue", "-")) if is_backend_available("cutile") else None,
    ("torch", "PyTorch", ("green", "-")),
```
Sorry for the inconvenience.

**Review by azazhu — APPROVED** [review]

**Comment by hannahli-nv** [comment]
/ok to test https://github.com/NVIDIA/TileGym/commit/d58112c13960a3677a2b29c343dde134ed4ce99d
